File: slides/UnionFind/union_find.cpp
#include <cstddef>
#include <cstdint>
#include <vector>
#include <cassert>

class union_find {
public:
    using size_type = std::size_t;
    using value_type = std::int_fast32_t;
protected:
    size_type _size;
    std::vector<value_type> _parent;
public:
    union_find() = default;
    union_find(size_type n): _size(n), _parent(n, value_type(-1)) { }

    void assign(size_type n) {
        _size = n;
        _parent.assign(n, value_type(-1));
    }

    void clear() {
        _size = 0;
        _parent.clear();
    }

    size_type root(size_type x) {
        assert(size_type(0) <= x and x < _parent.size());
        if (_parent[x] < 0) return x;
        return _parent[x] = static_cast<value_type>(root(_parent[x]));
    }

    size_type size() {
        return _size;
    }

    size_type size(size_type x) {
        assert(size_type(0) <= x and x < _parent.size());
        return static_cast<size_type>(-_parent[root(x)]);
    }

    bool same(size_type x, size_type y) {
        return root(x) == root(y);
    }

    bool unite(size_type x, size_type y) {
        size_type r_x = root(x), r_y = root(y);
        if (r_x == r_y) return false;
        if (_parent[r_x] > _parent[r_y]) std::swap(r_x, r_y);
        _parent[r_x] += _parent[r_y];
        _parent[r_y] = r_x;
        _size--;
        return true;
    }

};

File: slides/UnionFind/union_find.py
class union_find:
    __slots__ = ['_size', '_parent']
    def __init__ (self):
        self._size = 0
        self._parent = []

    def __init__ (self, n: int):
        self._size = n
        self._parent = [-1] * n

    def clear(self) -> None:
        self._size = 0
        self._parent.clear()

    def root(self, x: int) -> int:
        assert 0 <= x <= len(self._parent)
        if self._parent[x] < 0:
            return x
        else:
            self._parent[x] = self.root(self._parent[x])
            return self._parent[x]

    def size(self) -> int:
        return self._size

    def size(self, x) -> int:
        return -self._parent[x]
    
    def same(self, x: int, y: int) -> bool:
        return self.root(x) == self.root(y)
    
    def unite(self, x: int, y: int) -> bool:
        r_x, r_y = self.root(x), self.root(y)
        if r_x == r_y:
            return False
        if self._parent[r_x] > self._parent[r_y]:
            r_x, r_y = r_y, r_x
        self._parent[r_x] += self._parent[r_y]
        self._parent[r_y] = r_x
        self._size -= 1
        return True 